src/realm/group_writer.cpp

```diff
@@ -301,25 +302,283 @@ GroupWriter::MapWindow* GroupWriter::get_window(ref_type start_ref, size_t size)
+#if REALM_ALLOC_DEBUG
+        std::cout << " [" << start_pos << " - " << end_pos << "]";
+#endif
+        if (start_pos <= entry.ref && end_pos >= entry.ref + entry.size) {
+            // found a fully covering entry!
+            REALM_ASSERT(found_version <= entry.released_at_version);
+#if REALM_ALLOC_DEBUG
+            std::cout << "  backdating [" << entry.ref << ", " << entry.size
+                      << "]  version: " << entry.released_at_version << " -> " << found_version;
+#endif
+            return found_version;
+        }
+#if REALM_ALLOC_DEBUG
+        std::cout << "  not free at that point";
+#endif
+        return entry.released_at_version;
+    };
+
+
+    auto backdate_single_entry = [&](FreeSpaceEntry& entry, bool referenced) -> void {
+        static_cast<void>(referenced);
+        while (entry.released_at_version) {
+            // early out for references before oldest freelist:
+            if (entry.released_at_version <= this->m_oldest_reachable_version) {
+#ifdef REALM_DEBUG
+                REALM_ASSERT(!referenced);
+#endif
+                break;
+            }
+            auto earlier_it = get_earlier(entry.released_at_version);
+            // if no earlier versions than that exists, we're done
+            if (earlier_it == nullptr) {
+#if REALM_ALLOC_DEBUG
+                std::cout << "  backdating [" << entry.ref << ", " << entry.size << "] to zero (no earlier freelist)";
+#endif
+#ifdef REALM_DEBUG
+                REALM_ASSERT(!referenced);
+#endif
+                entry.released_at_version = 0;
+                break;
+            }
+            // earlier versions exist, we need to examine the freelist for it
+#if REALM_ALLOC_DEBUG
+            std::cout << " - earlier freelist: " << earlier_it->version;
+#endif
+            auto covering_blocks_released_at = find_cover_for(entry, *earlier_it);
+            if (covering_blocks_released_at == entry.released_at_version) {
+#ifdef REALM_DEBUG
+                REALM_ASSERT(referenced);
+#endif
+                break;
+            }
+#ifdef REALM_DEBUG
+            REALM_ASSERT(!referenced);
+#endif
+            entry.released_at_version = covering_blocks_released_at;
+        }
+    };
+
+
+#ifdef REALM_DEBUG
+    auto reachables = map_reachable();
+#if REALM_ALLOC_DEBUG
+    std::cout << "  Reachable: ";
+    ref_type last = 0;
+    for (auto& i : reachables) {
+        if (i.ref != last) {
+            last = i.ref;
+            std::cout << std::endl << "    " << i.ref << " - " << i.ref + i.size << " : " << i.released_at_version;
+        }
+        else {
+            std::cout << ", " << i.released_at_version;
+        }
+    }
+    std::cout << std::endl << "  Backdating:";
+#endif
+#endif
+    size_t limit = m_not_free_in_file.size();
+    for (size_t index = 0; index < limit; ++index) {
+        auto& entry = m_not_free_in_file[index];
+        bool referenced = false;
+#ifdef REALM_DEBUG
```

Comment:
Consider pulling this block out to an `is_referenced` lambda that always returns false unles `REALM_DEBUG` is active.